File: leetcode/Stack/leetcode-224/224.go
package stack

func calculate(s string) int {
	sigh, res := 1, 0
	stack := []int{}

	for i := 0; i < len(s); i++ {
		if s[i] >= '0' && s[i] <= '9' {
			num := 0
			for i < len(s) && s[i] >= '0' && s[i] <= '9' {
				num = num*10 + int(s[i]-'0')
				i++
			}
			res += sigh * num
			i--
		} else if s[i] == '+' {
			sigh = 1
		} else if s[i] == '-' {
			sigh = -1
		} else if s[i] == '(' {
			stack = append(stack, res)
			stack = append(stack, sigh)
			res = 0
			sigh = 1
		} else if s[i] == ')' {
			tmp_sigh := stack[len(stack)-1]
			tmp_res := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			res = res*tmp_sigh + tmp_res
		} else {
			continue
		}
	}
	return res
}

File: leetcode/Stack/leetcode-224/224.java
class Solution {
    public int calculate(String s) {
        if (s == null) {
            return 0;
        }
        int res = 0;

        int sigh = 1;
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                int num = 0;
                while (i < s.length() && Character.isDigit(s.charAt(i))) {
                    num = num * 10 + s.charAt(i) - '0';
                    i++;
                }
                res += sigh * num;
                i--;
            } else if (s.charAt(i) == '+') {
                sigh = 1;
            } else if (s.charAt(i) == '-') {
                sigh = -1;
            } else if (s.charAt(i) == '(') {
                stack.push(res);
                stack.push(sigh);
                res = 0;
                sigh = 1;
            } else if (s.charAt(i) == ')') {
                res = res * stack.pop() + stack.pop();
            } else {
                continue;
            }
        }
        return res;
    }
}